### DISCUSSION TITLE
mount /var/lib/docker/image to host for image persistency on delete?

### DISCUSSION THREAD

**Discussion by strobelight** [lima-vm/lima#1459] [Q&A] [answered]
I'm curious if `/var/lib/docker/image` directory can be mounted from the host in order that images don't disappear after a lima restart. Or, perhaps another solution for persistent images on lima restart using docker. I'm aware of the registry, but that's not used with `docker images` to my knowledge.

**Comment by afbjorklund** [comment]
What do you mean with "restart" ? All images should survive a simple stop + start of the instance, they are on the "diffdisk".

It would be possible to add a lima feature to cache images on the host, so that they could be loaded into new instances...

**Reply by jandubois** [reply]
> It would be possible to add a lima feature to cache images on the host, so that they could be loaded into new instances...

It doesn't need a new feature, you could simply mount the directory:

```yaml
mounts:
  location: ~/images
  mountPoint: /var/lib/docker/image
  writable: true
```

I haven't tested it, and it may be quite slow. I would also worry about stability.

A better solution would be to use `limactl disk create ...` to create another volume and then use a provisioning script to symlink `/var/lib/docker/image` to that location. Access would be fast, and Lima would make sure the volume isn't mounted to 2 different VMs at the same time.

**Reply by afbjorklund** [reply]
I meant something like `save` to and `load` from the cache directory on the host.

It would also be interesting to explore serving OCI directories and IPFS, on the host

**Reply by strobelight** [reply]
I was playing with getting my instance "just right", and do `limactl stop <instance>; limactl delete <instance>`, so no, not a simple restart.

**Reply by jandubois** [reply]
Yeah, if you delete the instance, all data is lost. You could have copied the directory to a shared host folder before deleting the instance, and then copying it back into the new instance.

But if you expect to regularly delete your instance, then storing persistent data in an additional volume is probably the best way.

**Reply by jandubois** [reply]
Note that you can add an additional disk at a later time; you just `limactl disk create ...`, then `limactl stop` your instance, and `limactl edit` to add it as an `additionalDisk`, and when you boot the instance again it shows up under `/mnt/lima-DISKNAME`.

**Reply by strobelight** [reply]
I'll play with `limactl disk create`; good info to check out.

Couldn't that all be put in the provisioning?

NM, I see `limactl edit` opens the same file I used to start it.

**Reply by jandubois** [reply]
> Couldn't that all be put in the provisioning?

The only thing that goes into a provisioning script would be to create a symlink or a bind-mount to the `/mnt/lima-DISKNAME` location because you cannot control the mount points for additional disks.

Something like (untested):

```yaml
additionalDisks:
- images

provision:
- mode: system
  script: |
    #!/bin/sh
    mkdir -p /var/lib/docker/image
    mount --bind /mnt/lima-images /var/lib/docker/image
```

But you have to create the disk (once) with e.g. `limactl disk create images --size 40G` before you configure it in a VM instance.

**Reply by strobelight** [reply]
I must not be using the correct directory; images always being pulled.

Do you happen to know the directory that images are pulled to so that I can link to my mount?

**Reply by AkihiroSuda** [reply]
`/var/lib/docker/*` has to be on a local filesystem such as ext4 or XFS.

**Reply by AkihiroSuda** [reply]
> so no, not a simple restart.

@strobelight Could you fix the OP to avoid confusion?

**Reply by afbjorklund** [reply]
Changed the title

**Reply by strobelight** [reply]
> /var/lib/docker/* has to be on a local filesystem such as ext4 or XFS.

@AkihiroSuda so the bind mount solution by @jandubois isn't a good one? So far, it seems to work using the same yaml file.

**Reply by jandubois** [reply]
The additional disk is formatted as `ext4` and is a local disk.

I think @AkihiroSuda was commenting that my first suggestion of using `sshfs` to mount the directory from the host is not going to work.

**Comment by strobelight** [comment]
I can't seem to get this extra disk mounted in user space.

**Reply by jandubois** [reply]
I'm not quite sure what you are trying to do. In rootless mode the data is stored under `$HOME/.local/docker`, so the provisioning script should be something like (again, untested):

```yaml
provision:
- mode: system
  script: |
    mkdir -p /home/${LIMA_CIDATA_USER}.linux/.local/docker
    mount --bind /mnt/lima-docker /home/${LIMA_CIDATA_USER}.linux/.local/docker
```

I would mount the full `docker/` directory and not just `docker/image` (and have created a `docker` disk to match this new usage). I'm not sure if just preserving the latter will be sufficient.

The provisioning script must still run as `root` because the user is not allowed to `mount` the directory.

**Reply by jandubois** [reply, marked as answer]
Finally got some time to actually test this, and here is a working version that persists both the root and user images on a separate volume:

```console
$ grep -A 18 additionalDisk ~/.lima/docker/lima.yaml
additionalDisks:
- docker

provision:
- mode: system
  script: |
    #!/bin/bash
    set -eux -o pipefail

    MNT=/mnt/lima-docker
    ROOT=/var/lib/docker
    USER=/home/${LIMA_CIDATA_USER}.linux/.local/share/docker

    mkdir -p $MNT/root $MNT/user
    mkdir -p $ROOT $USER
    mount --bind /$MNT/root $ROOT
    mount --bind /$MNT/user $USER

    chown -R $LIMA_CIDATA_USER:$LIMA_CIDATA_USER $USER

$ limactl shell docker ls -l /mnt/lima-docker
total 24
drwx------  2 root root 16384 Apr  6 01:56 lost+found
drwx--x--- 12 root root  4096 Apr  6 04:56 root
drwx--x--- 13 jan  jan   4096 Apr  6 04:55 user
```

Make sure you put the provision script first, before the other scripts already in `lima.yaml`!

<hr>

I just realized that the `chown` should probably start at `~/.local` if you start with a fresh VM, so something like

```
chown -R $LIMA_CIDATA_USER:$LIMA_CIDATA_USER /home/${LIMA_CIDATA_USER}.linux/.local
```

**Reply by strobelight** [reply]
I just want my images local and not ephemeral. I should be able to delete the instance, start a different one, mount the images, and not have to recreate them or pull them.

I was trying `{{.UID}}`; wasn't aware of `LIMA_CIDATA_USER` environment variable.

Thanks again!

**Reply by strobelight** [reply]
BTW, is the whole `docker` dir necessary to bind mount or can the `image` and `overlay2` be sufficient? Guess that might change depending on default storage driver.

**Reply by jandubois** [reply]
> is the whole `docker` dir necessary to bind mount

I don't know, so I thought it is better to play it safe. It doesn't hurt, as the bulk of the space used is by the images.

**Reply by afbjorklund** [reply]
> I just want my images local and not ephemeral.

The ubuntu VM is **not** ephemeral, there is one "pet" VM and it contains everything... It is possible to transform it to be more like the alpine VM, so that it has an "OS disk" and a "data disk" (like in the modified YAML), but that also makes the default more complex.

Maybe it should ask for confirmation, before deletion.

**Reply by strobelight** [reply]
> The ubuntu VM is not ephemeral

Even when instance stopped and deleted?

I discovered the images didn't stay around even in @jandubois solution if I put similar change into the rootful yaml example. Maybe I put the provision script in the wrong place, idk, but am happy with rootless solution as that'll be what I most use.

**Reply by jandubois** [reply]
> Even when instance stopped and deleted?

No, when the instance is deleted, then everything inside the VM is gone unless it is stored in an "additional disk".

I think @afbjorklund wanted to point out that in Alpine most of the OS is loaded from an ISO image into RAM (with the exception of `/etc /home /root /tmp /usr/local /var/lib`), so any changes to the OS are ephemeral: they need to be redone when you reboot the VM.

All other templates are starting with a disk image, not an ISO, and all changes made inside the VM are persisted across reboots. But not when you delete the instance.

> I discovered the images didn't stay around even in @jandubois solution if I put similar change into the rootful yaml example. Maybe I put the provision script in the wrong place, idk, but am happy with rootless solution as that'll be what I most use.

I can't tell why it didn't work without seeing exactly what you did, but if you are happy with what you have working, then maybe that is good enough too :)

**Reply by strobelight** [reply]
This is what I landed on: [docker-k3s-multiarch.yaml](https://github.com/strobelight/lima-docker-registry-k3s-persistent/blob/main/docker-k3s-multiarch.yaml).

After startup, also pulled other images, to confirm a stop/delete/start, kept them.

Then, stopped/deleted the instance and tried as rootful: [docker-k3s-multiarch-rootful.yaml](https://github.com/strobelight/lima-docker-registry-k3s-persistent/blob/main/docker-k3s-multiarch-rootful.yaml).

The extra container images weren't there.
